lib/logger/src/lib.rs

```diff
@@ -0,0 +1,102 @@
+#[macro_use]
+extern crate log;
+
+use std::cell::RefCell;
+use std::fs::File;
+use std::io::{self, Write};
+use std::sync::atomic;
+use std::thread;
+
+use log::{Level, LevelFilter, Metadata, Record};
+
+static STATE: atomic::AtomicIsize = atomic::ATOMIC_ISIZE_INIT;
+
+thread_local! {
+    static WRITER: RefCell<Option<io::BufWriter<File>>> = RefCell::new(None);
+}
+
+pub fn initialize(filename_prefix: &str, enable: bool) {
+    // Ensure the thread local state is always properly initialized.
+    WRITER.with(|rc| {
+        if rc.borrow().is_none() {
+            rc.replace(Some(open_file(filename_prefix)));
+        }
+    });
+
+    let new_state = match enable {
+        true => 1,
+        false => -1,
+    };
+
+    let previous_state = STATE.load(atomic::Ordering::Relaxed);
+    if previous_state == new_state {
+        return;
+    }
+
+    // Once and only once in the entire program, set up the logger.
+    if previous_state == 0 {
+        log::set_boxed_logger(Box::new(ThreadPerFileLogger))
+            .map(|()| log::set_max_level(LevelFilter::Info))
+            .expect("setting up logging panicked");
+    }
+
+    STATE.store(new_state, atomic::Ordering::Relaxed);
+    info!(
+        "Set up logging to {:?}; filename prefix is {}",
+        enable, filename_prefix
+    );
+}
+
+#[inline]
+fn enabled() -> bool {
+    STATE.load(atomic::Ordering::Relaxed) > 0
+}
+
+struct ThreadPerFileLogger;
+
+impl log::Log for ThreadPerFileLogger {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() <= Level::Info && enabled()
```

Comment:
I have to admit this is just a copy/paste of the `log` example. In fact, this will allow up to `Info` but the order is inverted, so we'd have `error!`, `warn!` and `info!` show up, but not `debug!` and `trace!`. Will set to `trace` to allow every type of log commands at the moment, maybe we can add a way to control via an environment variable.